mavros/src/plugins/wind_estimation.cpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * @brief wind estimation plugin
+ * @file wind_estimation.cpp
+ * @author Thomas Stastny <thomas.stastny@mavt.ethz.ch>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2018 Thomas Stastny <thomas.stastny@mavt.ethz.ch>
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <angles/angles.h>
+#include <mavros/mavros_plugin.h>
+
+#include <geometry_msgs/TwistWithCovarianceStamped.h>
+
+namespace mavros {
+namespace std_plugins {
+/**
+ * @brief Wind estimation plugin.
+ */
+class WindEstimationPlugin : public plugin::PluginBase {
+public:
+	WindEstimationPlugin() : PluginBase(),
+		nh("~")
+	{ }
+
+	/**
+	 * Plugin initializer. Constructor should not do this.
+	 */
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		wind_pub = nh.advertise<geometry_msgs::TwistWithCovarianceStamped>("wind_estimation", 10);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return {
+			make_handler(&WindEstimationPlugin::handle_APM_wind),
+			make_handler(&WindEstimationPlugin::handle_PX4_wind),
+		};
+	}
+
+private:
+	ros::NodeHandle nh;
+
+	ros::Publisher wind_pub;
+
+	/**
+	 * Handle APM specific wind direction estimation message
+	 */
+	void handle_APM_wind(const mavlink::mavlink_message_t *msg, mavlink::ardupilotmega::msg::WIND &wind)
+	{
+		const double speed = wind.speed;
+		const double course = angles::from_degrees(wind.direction);
+
+		auto twist_cov = boost::make_shared<geometry_msgs::TwistWithCovarianceStamped>();
+		twist_cov->header.stamp = ros::Time::now();
+		// TODO: check math's
+		twist_cov->twist.twist.linear.x = speed * std::sin(course);
```

Comment:
No frame transforms?

Reply:
this is a copy paste of the APM handler that was in VFR_HUD before. I assume this should be ok? Unless someone with APM experience could point out any changes since that handler was written.